### PR TITLE
add JSON playlist format

### PR DESCRIPTION
**Pull request by bparker06**
## Description

This adds a new playlist format, on by default, using JSON. Example syntax:

```
[
  {
    "path": "/home/bp/roms/sfc/Star Fox (J).smc",
    "label": "Star Fox (Japan)",
    "core_path": "DETECT",
    "core_name": "DETECT",
    "crc32": "41A60B3F|crc",
    "db_name": "Nintendo - Super Nintendo Entertainment System.lpl"
  }
]
```

This allows us the flexibility to more easily add support for new things such as [per-content playtime logs](https://github.com/libretro/RetroArch/issues/3527), subsystem info or other metadata.

The old format is still fully supported and the file extension has not changed. Users are not forced to use this new format and may opt-out of it if they wish.

There is a new option under Settings -> Playlist to change the format used for writing out any new/updated playlists. The default is currently the new format. This means that in addition to newly created playlists, old playlists that are updated in any way (entries added or deleted) will be converted to the new format.

When reading playlists in, the correct format is detected regardless of the above setting. The setting only changes the format used when files are written to disk.

## Related Issues

#3527
#5360

## Reviewers

@twinaphex 
@fr500 
@hunterk 
@Tatsuya79

### CODE REVIEW

**Comment by markwkidd** [comment]
Awesome!! To me, the main thing is that we have a format that is extensible which this would seem to be.

**Comment by hizzlekizzle** [comment]
this looks/sounds great to me, and it opens the door for lots of other cool stuff in the future. Great job, @bparker06

**Comment by RobLoach** [comment]
70b

**Comment by Tatsuya79** [comment]
It's working.
But after conversion my Mame playlist of 16000+ entries went from 1.37MB to 3.16MB.

So we could have additional infos in playlists like "manufacturer", "year" and display them as new columns in QT UI playlist detail view?

**Comment by i30817** [comment]
Could we have a way to reset core association with this (finally?)

**Comment by markwkidd** [comment]
@i30817 reset core association does work with the old playlist format. just making sure you know this

**Comment by i30817** [comment]
Does it really... because i can't find it anywhere both by selecting the game on the playlist or selecting the game on the history tab 'playlist'.

**Comment by markwkidd** [comment]
@i30817 selecting the game on the playlist shows the "Reset Core Association" option for me


screenshot 2019-01-10 13 07 50

**Comment by i30817** [comment]
Doesn't exist in xmb.

**Comment by i30817** [comment]
In fact, i can't find it on rgui either (and yes, i'm running a game i ran before).

Maybe it's because i associated the playlist to a core on the playlist options, but i just tried to reset the playlist to N/A on the game console i ran before and even then i can't change (and yes, i do have 2 cores for that console, mupen64 and parellel64).

**Comment by markwkidd** [comment]
In XMB:

screenshot 2019-01-10 14 30 54

**Comment by i30817** [comment]
Yeah, that doesn't happen. And i checked the files and the playlist has a core association with the game i'm trying:
```
/media/i30817/Games/Nintendo 64/Banjo-Kazooie (USA) (Rev A)/Banjo-Kazooie (USA) (Rev A).n64
Banjo-Kazooie (USA) (Rev A)
/home/i30817/.config/retroarch/cores/parallel_n64_libretro.so
Nintendo - Nintendo 64 (ParaLLEl N64)
90B38651|crc
Nintendo - Nintendo 64.lpl
```
and the setting is turned on in the cfg:  `quick_menu_show_reset_core_association = "true"`

But as mentioned i'm going to wait until flatpak updates before making a stink about it.

**Comment by RobLoach** [comment]
> @i30817 But as mentioned i'm going to wait until flatpak updates before making a stink about it.

Flatpak will always only run the latest stable. So, when 1.7.6 is out, you may be able use this if it gets merged. To test the JSON playlist right now, you'll want to checkout bparker's branch and compile it yourself...

```
git clone https://github.com/bparker06/RetroArch.git
cd RetroArch
git checkout json_playlist
./configure
make
./retroarch
```

**Comment by RobLoach** [comment]
This is working well! Some feedback...

1. While the file contents' is saved as JSON format, the file extension is listed as a `.lpl`. Should we consider saving/loading the new format files as `.json` to differentiate between the two?

2. Could likely be a follow up, but it may be helpful to provide a sublabel? "When enabled, will save files as the legacy .lpl format. Otherwise, will save files in JSON format. Recommended to be disabled."
oldformat

**Comment by i30817** [comment]
But why keep the old format? Just have a conversion procedure if it finds lpl files and delete those sounds saner than giving a superfluous choice that will almost immediately turn into technical debt once options are implemented in json that aren't on the original format.

**Comment by hizzlekizzle** [comment]
I don't think it needs the json file format, personally. However, I was thinking we should probably plan to slowly deprecate the old format over the next few versions. For example, next version we default to new and optionally change old to new on modification (i.e., current PR behavior), next version we remove the option to create old playlists and change old to new on access. After that, we could potentially remove the old parsing altogether, though I don't suppose it really hurts anything to keep it around.

**Comment by fr500** [comment]
It depends.
Does 3DS/VITA/PSP support the new format?
Remember old consoles are memory contrained, I'd keep the old format for
psp/xbox/wii if at all possible.

Actually I'd remove scanning and playlist on PSP at least :P

**Comment by bparker06** [comment]
>So we could have additional infos in playlists like "manufacturer", "year" and display them as new columns in QT UI playlist detail view?

Yes you could technically do something like that.

>I don't think it needs the json file format, personally. 

Why? What format do you think would be better? You previously suggested JSON yourself [here](https://forums.libretro.com/t/proposal-for-playlist-file-format-v2/14667/9). Also don't you think we have enough parsers already?

>Does 3DS/VITA/PSP support the new format

Griffin builds already include the JSON reader/writer if HAVE_NETWORKING is defined, so it will work in those cases (guess it should be moved out of there for this PR since JSON would no longer be used only for netplay). And for Makefile.common platforms, it's already enabled no matter what.

>I'd keep the old format for psp/xbox/wii if at all possible

That's easily done with a platform ifdef to force the default setting for the new option, if that's what people want to do.

**Comment by hizzlekizzle** [comment]
>Why? What format do you think would be better?
no, no, I was replying to RobLoach's suggestion that we change file extension to *.json. I indeed think json syntax is the way to go.

**Comment by Ferk** [comment]
One question that has always bugged me about the playlist format: why does the db_name need to be stored for every single entry? isn't it meant to be the same for the same file?

Maybe instead of being an array it could be an object so the playlist itself could have metadata.
Or possibly have it store default values for all the entries in the list. This way the size of the archives could be significantly reduced.

{code}
{
  "db_name": "Nintendo - Super Nintendo Entertainment System.lpl",
  "core_path": "DETECT",
  "core_name": "DETECT",
  "entries": [
	  {
		"path": "/home/bp/roms/sfc/Star Fox (J).smc",
		"label": "Star Fox (Japan)",
		"crc32": "41A60B3F|crc",
	  },
	  {
		"path": "/home/bp/roms/sfc/My Awesome game.smc",
		"label": "Amazeballs!",
		"crc32": "DETECT",
	  }
	]
}
{code}

**Comment by i30817** [comment]
> One question that has always bugged me about the playlist format: why does the db_name need to be stored for every single entry? isn't it meant to be the same in the entire file?

I don't think this is necessarily true. For instance what if some day, for some reason (such as different verification methods required), RA wants to distinguish snes hacks from snes games in the same playlist? It can do so from different db_names. Though that is a distinction that is more useful before scanning.

**Comment by bparker06** [comment]
@Ferk,

>isn't it meant to be the same in the entire file

In practice, no, and especially with the new desktop menu's playlist editor, it's quite easy to add entries referencing different databases in the same playlist.

For example, some users might have a "Mario" playlist, with Mario games from all different systems. The db_name field is then also used as the folder name to look up the thumbnails (at least in the desktop menu), so for playlists with entries for different systems (or say, hacks combined with normal roms into one playlist, since they're different databases), this becomes more important.

**Comment by hizzlekizzle** [comment]
Up to you, I reckon @bparker06. I just didn't want this PR to get caught in implementation committee hell.

Re: version number, yes! That is an excellent idea. Is there any other parent metadata we need to get in there from the get-go, so we don't fall into the GLSL version directive trap?

**Comment by bparker06** [comment]
I can't think of anything at the moment, but I guess it can always be added later, especially since we have a version number to go by. I'll go ahead and do the parent node thing and push it here.

**Comment by bparker06** [comment]
Done, now it looks like this:

```
{
  "version": "1.0",
  "items": [
    {
      "path": "/home/bp/roms/DC/Bangai-O (USA).gdi",
      "label": "Bangai-O (USA)",
      "core_path": "DETECT",
      "core_name": "DETECT",
      "crc32": "B0D7F52E|crc",
      "db_name": "Sega - Dreamcast.lpl"
    }
  ]
}
```

**Comment by bparker06** [comment]
@twinaphex Is there anything else that needs to be done for this?

File: playlist.c [hunk 1]
```diff
@@ -491,66 +642,336 @@ size_t playlist_size(playlist_t *playlist)
...
+   {
+      char buf[16] = {0};
+      int64_t bytes_read = filestream_read(file, buf, 15);
+
+      filestream_seek(file, 0, SEEK_SET);
 
-   for (playlist->size = 0; playlist->size < playlist->cap; )
+      if (bytes_read == 15)
+      {
+         if (string_is_equal(buf, "{\n  \"version\": "))
```
Comment (RobLoach) [inline playlist.c]: The detection between the differents formats while using the same file extension is the only issue I have with this. Shouldn't be that much of a problem though, as long as we make sure `{\n  "version": ` is always first in the new playlists' contents.

Leaving it to @twinaphex to do the final review! Nice work, bparker06. Excited to see how we can expand the playlist functionality through this.

**Pull request merged**